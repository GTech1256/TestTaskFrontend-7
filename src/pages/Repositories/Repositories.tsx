import { useQuery } from '@apollo/client';
import React, { useCallback, useState } from 'react';

import RepositoriesLicenseSelect from './containers/RepositoriesLicenseSelect';
import RepositoriesView from './components/RepositoriesView';
import {
  AllRepositoriesQueryType,
  getRepositiryQueryString,
  ALL_REPOSITORIES,
  MIN_STARS_COUNT,
} from '../../services/graphql/queries/getRepositories';
import Input, { InputProps } from '../../components/Input';
import { SelectProps } from '../../components/Select';
import Pagination from '../../components/Pagination';
import withDebounce from '../../hoc/withDebounce';


const InputWithDebounce = withDebounce(Input)
const RepositoriesLicenseSelectWithDebounce = withDebounce(RepositoriesLicenseSelect)

const Repositories = () => {
  const [license, setLicense] = useState<string | undefined>(undefined)
  const [name, setName] = useState<string | undefined>(undefined)
  const [beforeCursor, setBeforeCursor] = useState<string | undefined>(undefined)
  const [afterCursor, setAfterCursor] = useState<string | undefined>(undefined)
  
  const { loading, error, data } = useQuery<AllRepositoriesQueryType>(ALL_REPOSITORIES, {
    variables: {
      query: getRepositiryQueryString({license, name }),
      after: afterCursor,
      before: beforeCursor
    },
  });

  const resetPagination = useCallback(() => {
    setBeforeCursor(undefined)
    setAfterCursor(undefined)
  }, [])

  const handleLicenseSelect = useCallback<SelectProps['onChange']>((value) => {
    resetPagination()
    setLicense(value)
  }, [resetPagination])
  const handleNameChange = useCallback<InputProps['onChange']>((value) => {
    resetPagination()
    setName(value)
  }, [resetPagination])
  const handlePaginationNext = useCallback(() => {
    setBeforeCursor('' + data!.search.pageInfo.endCursor)
    setAfterCursor(undefined)
  }, [data])
  const handlePaginationPrev = useCallback(() => {
    setBeforeCursor(undefined)
    setAfterCursor('' + data!.search.pageInfo.startCursor)
  }, [data])

  return (
    <div>
      <div className="alert alert-primary" role="alert">
        Репозитории с менее чем {MIN_STARS_COUNT} звезд не будут найдены из-за &nbsp;
        <a
          href="https://docs.github.com/en/github/searching-for-information-on-github/troubleshooting-search-queries#potential-timeouts"
          target="_blank"
          rel="noopener noreferrer"
        >ограничений GitHub</a>
      </div>
      <RepositoriesLicenseSelectWithDebounce
        value={license}
        onChange={handleLicenseSelect}
        name="repositoriesLicenseSelect"
      />
      <InputWithDebounce
        name={'repositoryName'}
        onChange={handleNameChange}
        placeholder="Введите имя репозитория"
        label="Имя репозитория"
      />
      <Pagination
        hasNextPage={data?.search.pageInfo.hasNextPage}
        hasPreviousPage={data?.search.pageInfo.hasPreviousPage}
        onNext={handlePaginationNext}
        onPrev={handlePaginationPrev}
      />
      {
        loading ? <p className="text-center">Загрузка Репозиториев...</p> :
        error ? <p className="text-center">Не удалось загрузить репозиториии</p> :
        data?.search.edges.length === 0 ?
          <p>Репозиториев по такому фильтру не найдено</p> :
          <RepositoriesView className="mb-5" reposetories={data!.search.edges} />
      }
    </div>
  )
}

export default Repositories;