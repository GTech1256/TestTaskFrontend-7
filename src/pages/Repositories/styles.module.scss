.repositories {
  display: flex;
  flex-direction: column;
}

.repositories__text {
  margin: auto 0;
}