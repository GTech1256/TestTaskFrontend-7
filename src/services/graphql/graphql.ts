import { setContext } from '@apollo/client/link/context';
import { ApolloClient, createHttpLink, InMemoryCache } from '@apollo/client';

import { getGithubToken } from '../github';


const httpLink = createHttpLink({
  uri: 'https://api.github.com/graphql',
});

const authLink = setContext(async (_, { headers }) => {
  const token = await getGithubToken();

  // return the headers to the context so httpLink can read them
  return {
    headers: {
      ...headers,
      authorization: token ? `Bearer ${token}` : "",
    }
  }
});

export const client = new ApolloClient({
  link: authLink.concat(httpLink),
  cache: new InMemoryCache()
});