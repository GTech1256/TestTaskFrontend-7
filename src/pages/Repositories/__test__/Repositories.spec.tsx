import React from 'react';
import renderer from 'react-test-renderer';
import { ApolloProvider } from '@apollo/client';

import { client } from '../../../services/graphql';
import Repositories from '../Repositories';


describe('<Repositories />', () => {
  it('should render correct', () => {
    const tree = renderer
      .create(<ApolloProvider client={client}><Repositories /></ApolloProvider>)
      .toJSON();

    expect(tree).toMatchSnapshot();
  })
})