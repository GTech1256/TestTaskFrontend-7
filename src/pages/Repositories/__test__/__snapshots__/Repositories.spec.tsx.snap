// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Repositories /> should render correct 1`] = `
<div>
  <div
    className="alert alert-primary"
    role="alert"
  >
    Репозитории с менее чем 
    1600
     звезд не будут найдены из-за  
    <a
      href="https://docs.github.com/en/github/searching-for-information-on-github/troubleshooting-search-queries#potential-timeouts"
      rel="noopener noreferrer"
      target="_blank"
    >
      ограничений GitHub
    </a>
  </div>
  <div
    className="input-group mb-3 "
  >
    <div
      className="input-group-prepend"
    >
      <label
        className="input-group-text"
        htmlFor="repositoriesLicenseSelect"
      >
        Лицензия
      </label>
    </div>
    <select
      className="custom-select"
      defaultValue="DEFAULT"
      disabled={true}
      id="repositoriesLicenseSelect"
      onChange={[Function]}
    >
      <option
        value="DEFAULT"
      >
        Фильтр по лицензии...
      </option>
    </select>
  </div>
  <div
    className="input-group mb-3 "
  >
    <div
      className="input-group-prepend"
    >
      <label
        className="input-group-text"
        htmlFor="repositoryName"
      >
        Имя репозитория
      </label>
    </div>
    <input
      aria-describedby="basic-addon1"
      aria-label="repositoryName"
      className="form-control"
      name="repositoryName"
      onChange={[Function]}
      placeholder="Введите имя репозитория"
    />
  </div>
  <nav
    aria-label="Page navigation example"
  >
    <ul
      className="pagination justify-content-center"
    >
      <li
        className="page-item disabled"
      >
        <button
          className="page-link"
          disabled={true}
          onClick={[Function]}
        >
          <span
            aria-hidden="true"
          >
            «
          </span>
          <span
            className="sr-only"
          >
            Предыдущая
          </span>
        </button>
      </li>
      <li
        className="page-item disabled"
      >
        <button
          className="page-link"
          disabled={true}
          onClick={[Function]}
        >
          <span
            aria-hidden="true"
          >
            »
          </span>
          <span
            className="sr-only"
          >
            Следующая
          </span>
        </button>
      </li>
    </ul>
  </nav>
  <p
    className="text-center"
  >
    Загрузка Репозиториев...
  </p>
</div>
`;
